provider/provider_test.go

```diff
@@ -153,3 +78,27 @@ func pulumiTest(t *testing.T, dir string, opts ...opttest.Option) *pulumitest.Pu
 	test.SetConfig(t, "gcp:config:project", googleProj)
 	return test
 }
+
+func testUpgrade(t *testing.T, dir string, opts ...optproviderupgrade.PreviewProviderUpgradeOpt) auto.PreviewResult {
+	t.Helper()
+	if testing.Short() {
+		t.Skipf("Skipping in testing.Short() mode, assuming this is a CI run without cloud credentials")
+		return auto.PreviewResult{}
+	}
+
+	rpFactory := providers.ResourceProviderFactory(providerFactory)
```

Comment:
non-blocking: what is the purpose of a `ResourceProviderFactory`? The [library docs](https://github.com/pulumi/providertest/blob/387f6b5565e7d34dc2314b9b12b8974894038a9e/providers/resourceProvider.go#L14) seem to describe a different function entirely.

More specifically to this PR: can we add a quick comment elaborating the purpose of this providerFactory (as opposed to `opttest.LocalProvider()`) - what does it allow the test framework to do?

Reply:
It allows us to use an in-process provider which automatically gets rebuilt for tests addressing one of your points from slack about tests not running against your latest changes.